### ISSUE TITLE
Support PackageReference system, e.g. .csproj.

### ISSUE THREAD

**Issue by jwittner** [GlitchEnzo/NuGetForUnity#88]
Would be great to support the latest and greatest package reference systems!

Documentation for [PackageReference](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files)

**Comment by mhutch** [comment]
I think a pretty good way to handle this would be to replace `packages.config` with a `packages.props` file that contained just the PackageReference items.

For NuGetForUnity, it would simply be a package manifest, however the syntax would be the same as PackageReference in csproj files, which would. be good for consistency. The csproj that's generated for Unity Tools could also `<Import>` it directly.

**Comment by GlitchEnzo** [comment]
I haven't yet had the pleasure to use VS 2017, so I don't know much about this feature.

Since Unity automatically re-generates the .sln and .csproj files whenever any code changes in the project, these tweaks would be lost.  As @mhutch suggests, an intermediate file such as `packages.props` file would have to exist and we'd have to inject its contents into the .csproj files using AssetPostProcessor.OnGeneratedCSProjectFile() or some similar mechanism.

Pardon my ignorance, but what is the advantage of switching to use PackageReferences in the .csproj file via an intermediate file (which would only work in VS 2017+) versus the current `packages.config` approach (which works in VS 2012+)?

**Comment by mhutch** [comment]
https://blog.nuget.org/20170316/NuGet-now-fully-integrated-into-MSBuild.html

tl;dr faster install/restore, system-wide package cache (instead of solution-local), assets injected directly into build instead of modifying project, msbuild-integrated restore command, transitive dependencies (no need to list entire closure of package dependencies in manifest), ability to control over which assets are consumed

Not all of these would apply to Unity, but I think consistency would be good.

**Comment by sailro** [comment]
cc @jbevain

Currently all csproj+sln files are generated either by Unity (if you use MonoDevelop/Rider/VSCode/misc) or the VSTU plugin (if VSTU is installed + you selected Visual Studio -including VS4Mac- as an external editor). So one option is to directly handle this in our file generator instead of injecting/manipulating generated files. 

Seb
Visual Studio Tools for Unity

**Comment by GlitchEnzo** [comment]
@mhutch Thanks for the link! Transitive Package Restore alone is a _huge_ benefit. I've done a lot of work optimizing the package restore operation as much as possible, so I welcome anything that can help speed that up.

@sailro VSTU having "native" support for NuGet in the .csproj and .sln generation would be awesome!

**Comment by jwittner** [comment]
@GlitchEnzo  - I read @mhutch recommendation as replacing our .config file with a .props file instead of leveraging the .csproj generated by Unity. I like this as it maintains support fully within Unity. VSTU could then import those props into the solution.

**Comment by jwittner** [comment]
One thing I like about the intermediate props file is that it allows support for NuGet in Unity without requiring Visual Studio or the VSTU plugin. I think this is a better story for NuGet and for Unity. 

VSTU automatically referencing those props files on generation seems to wrap it all up in a nice bow when you opt-in to Visual Studio.

**Comment by GlitchEnzo** [comment]
As is it right now, with the `packages.config` intermediate file, NuGetForUnity is IDE-agnostic.  It doesn't matter what version of Visual Studio they are using, or if they are even using VS at all.

Having VSTU `<Import>` the .props file would be an awesome bonus, but it would _only_ work where the user (A) has Visual Studio 2017 or newer installed and (B) has VSTU installed.

I'm in full support of switching the intermediate file from `packages.config` to `packages.props` as long as we maintain the current IDE-agnostic behavior.

**Comment by sailro** [comment]
We know which VS version is currently used as an external editor in Unity (VSTU is installed per specific VS version). So if we are under VS2017+ and we have this nuget.props file we could import the file, and fallback to the legacy generation with VS2015 and lower. If VS is not the external editor it will use the default Unity generator so everything should work as expected with all IDEs.

But in the end it is a bit more complicated than just importing the prop file, as we have to stop directly referencing assemblies+dependencies (we currently use Reference Include=... for all) as PackageReference will now do this job.

**Comment by CDTR-MattConroy** [comment]
6 years on, is there any advances on this?
It's a shame Unity isn't yet using SDK style projects, but it would be great to utilise PackageReference nonetheless.

**Comment by JoC0de** [comment]
Just to clear thinks up, NuGetForUnity "just" downloads NuGet packages and places the .dll into the Unity project (simplified 😃). The `packages.config` file is just a file where NuGetForUnity stores the information of what packages to install, this could be any file / file format. Unity has a complet different build system, not supporting anything the "normal" .Net build system (MSBuild) supports. Unity has a package [Visual Studio Editor](https://docs.unity3d.com/Packages/com.unity.ide.visualstudio@2.0/manual/index.html) to generate `.sln` and `.csproj` files. This files can only be used to open the project in a IDE they are not used by Unity when it builds / generates the executable for your App / Game. So switching to `PackageReference` would change nothing because Unity still needs the .dll to includes the package assembly in the build. The only think we probably can do is use the new cache location where NuGet packages installed with  `PackageReference` are located instead of using the old NuGet package cache directory.
Unity itself already works on supporting the "normal" .Net tooling MSBuild and the newer .Net versions 5+ so it will get official NuGet support. There is a interesting [Blog post ](https://blog.unity.com/engine-platform/unity-and-net-whats-next) and a constantly updated [Forum thread](https://forum.unity.com/threads/unity-future-net-development-status.1092205) tracking the progress.

**Comment by CDTR-MattConroy** [comment]
Thanks, this makes sense. The big change I was hoping for was transitive package dependencies. My NuGet dependencies are quite simple at the moment, but in the past I quickly ran in dependency nightmares between packages.config and the binding redirects.

Interesting that you mention the analysers too as I do get issues the first time I install a package with analysers, for example AWS packages. It complains the references can't be validated and I have to restart Unity before it will compile.

Is the bottom line that we need to wait for Unity to update their build pipeline and support SDK-style projects?

**Comment by JoC0de** [comment]
> Thanks, this makes sense. The big change I was hoping for was transitive package dependencies. My NuGet dependencies are quite simple at the moment, but in the past I quickly ran in dependency nightmares between packages.config and the binding redirects.

What do you mean by 'transitive package dependencies'? Do you mean that in `PackageReference` the .dll is only placed inside the  'cache' and not copied to the project directory? NuGetForUnity should handle dependencies of NuGet packages, yes in some cases it can get a little bit bloated. In cases where you build a normal .net project alongside the Unity project it could be that using the newer cache location could remove the need of downloading a NuGet package twice. But downloading is only a one time thing, than it is cached anyways.

Binding redirects aren't even supported by Unity (as far as I know) so you can only ignore the warnings about 'version conflicts', you probably get in Visual Studio or even sometimes in Unity.

> Interesting that you mention the analysers too as I do get issues the first time I install a package with analysers, for example AWS packages. It complains the references can't be validated and I have to restart Unity before it will compile.

Installing NuGet packages that contain analyzers is supported by NuGetForUnity (in newer version) and you even get the warnings inside the Unity Editor.

> Is the bottom line that we need to wait for Unity to update their build pipeline and support SDK-style projects?

In the Forum / Blog Unity only mentions updating to new .Net version they don't provide information if they still relay on `.asmdef`  or if thy switch to `.csproj` files. See e.g. the recent comment:
![image](https://github.com/GlitchEnzo/NuGetForUnity/assets/53140583/b4675fc0-a13f-4182-9f14-dd8daa01d6b3)
https://forum.unity.com/threads/unity-future-net-development-status.1092205/page-43#post-9386252

For NuGet packages that e.g. rely on a .net version > 5 we more or less rely on Unity.

**Comment by CDTR-MattConroy** [comment]
> What do you mean by 'transitive package dependencies'?

Transitive package dependencies are when installing a package will bring its dependencies with them but they are not themselves "installed". 

For example if I installed Serilog.Sinks.File, one of its dependencies is Serilog. I don't need to install Serilog directly but I do require it's DLL as Serilog.Sinks.File depends on it. Installing it as a top level dependency clutters the packages.config and I now can't be sure if packages are installed because I use them or because something else does. It also means I have to pick a version to install rather than letting Serilog.Sinks.File have whatever version it wants, and whatever that version may or may not change to in future.

PackageReferences use transitive dependencies, meaning that any top level package I install will bring its dependencies with it. I'm free to install their dependencies too, if I want to use them explicitly or specific a version higher than what is required. But it means I only have the packages I actually require specified and if I uninstall one that is a dependency of another it becomes a transitive dependency once again and the DLL will still be included.

> Binding redirects aren't even supported by Unity (as far as I know) so you can only ignore the warnings about 'version conflicts', you probably get in Visual Studio or even sometimes in Unity.

Luckily I've not fallen afoul of version conflicts yet. But as projects grow larger it becomes more difficult to maintain a set of packages that work together. In the past I've often gotten to the point where installing new ones or updating existing ones becomes dangerous as you're never sure if you've affected one of your other packages dependencies. Again transitive dependencies help here as each top level packages always knows it's dependencies so you can be aware of any issues of conflicts at install time rather than run time!


> Installing NuGet packages that contain analyzers is supported by NuGetForUnity (in newer version) and you even get the warnings inside the Unity Editor.

I'm currently on 3.1.3, which was the latest last I checked. I've had the error on that version. I'll take some details and open an issue when it occurs again.


> In the Forum / Blog Unity only mentions updating to new .Net version they don't provide information if they still relay on `.asmdef`  or if thy switch to `.csproj` files. See e.g. the recent comment:
> 
> ![image](https://github.com/GlitchEnzo/NuGetForUnity/assets/53140583/b4675fc0-a13f-4182-9f14-dd8daa01d6b3)
> 
> https://forum.unity.com/threads/unity-future-net-development-status.1092205/page-43#post-9386252
> 
> 
> 
> For NuGet packages that e.g. rely on a .net version > 5 we more or less rely on Unity.

I've been following that thread for a while and unfortunately it still sounds like we're a few years away from having proper .csproj support in Unity. Including newer .NET versions, C# versions and things like PackageReferecnes.

**Comment by JoC0de** [comment]
Oh I really should publish a newer release because the changes in #539 should solve your transitive dependency problem. I expected this to already be in the latest Release. With the change the UI looks something like in the [screenshot](https://github.com/GlitchEnzo/NuGetForUnity/blob/master/docs/screenshots/installed.png). So when you update a package you know if it is only a transitive dependency or installed manually. In the `package.config` the transitive dependency are still listed but only to have a shortcut to know what packages are installed / should be installed. So we treat the `package.config` more like a `package-lock.json` in NPM. Theoretically we can even remove the entries from the `package.config` file / fully hide them from the UI but it would make some thinks a little bit harder and even Visual Studio now shows transitive dependencies inside the NuGet package window.

**Comment by CDTR-MattConroy** [comment]
Interesting, what does the resulting `packages.config` look like? That sounds like it's the same as nuget's transitive dependencies, but you've implemented them yourself. I'll make sure to update when I see a new version 👍

**Comment by igor84** [comment]
This is how it looks like after installing Grpc.Core package:
```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Grpc.Core" version="2.46.6" manuallyInstalled="true" />
  <package id="Grpc.Core.Api" version="2.46.6" />
</packages>
```
As you can see the Grpc.Core has a manuallyInstalled="true" attribute while its dependency does not. If I were to click uninstall on Grpc.Core it would in this case uninstall both packages because there is no other package that depends on Grpc.Core.Api.

**Comment by CDTR-MattConroy** [comment]
Coming back to this after some time. I have found that these latest changes go some way towards solving the issue, but still mean that I would be able to uninstall a package's dependencies which, IMO, shouldn't be allowed. I should be able to not "manually install" it, but it must still hang around if another package is dependent on it. Which ultimately means I'm having to uninstall packages when I want to no longer references one of it's dependencies, in order to re-install it afterwards and bring back its dependencies.

The way PackageReference works in standard .NET projects means that I only specify the packages I want to reference directly, and their dependencies are resolved and restored independently, meaning that if `Grpc.Core` relies on `Grpc.Core.Api`, I only need to reference `Grpc.Core`, and if another project relies on a different version of `Grpc.Core.Api`, then the lowest appropriate version is restored, or else package restore fails, as opposed to run-time errors when DLLs are missing.

I hope that makes sense, and would be great to hear your thoughts.
